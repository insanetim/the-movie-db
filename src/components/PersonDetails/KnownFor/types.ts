import { ICredit, IPersonDetails } from 'src/interfaces/person.interface'

export type KnownForProps = {
  credits: IPersonDetails['movie_credits']
  knownForDepartment: IPersonDetails['known_for_department']
}

export type KnownForHookProps = {
  credits: IPersonDetails['movie_credits']
  knownForDepartment: IPersonDetails['known_for_department']
}

export type KnownForHookReturn = {
  sortedItems: ICredit[]
}
