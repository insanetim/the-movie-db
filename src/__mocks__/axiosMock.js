function mockAxios() {
  const request = jest.fn()
  const mockedInterceptorsManager = {
    use: jest.fn(() => ({ type: 'interceptor-usage-token' })),
    eject: jest.fn()
  }
  const mockedProxyClient = Object.assign(request, {
    request,
    interceptors: {
      request: mockedInterceptorsManager,
      response: mockedInterceptorsManager
    },
    get: jest.fn((url, config) => request({ ...config, url, method: 'get' })),
    delete: jest.fn((url, config) => request({ ...config, url, method: 'delete' })),
    head: jest.fn((url, config) => request({ ...config, url, method: 'head' })),
    options: jest.fn((url, config) => request({ ...config, url, method: 'options' })),
    post: jest.fn((url, config) => request({ ...config, url, method: 'post' })),
    put: jest.fn((url, data, config) => request({ ...config, url, method: 'put', data })),
    patch: jest.fn((url, config) => request({ ...config, url, method: 'patch' }))
  })

  return mockedProxyClient
}

export default mockAxios
