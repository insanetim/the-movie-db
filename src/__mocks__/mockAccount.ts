import { IAccount } from 'src/interfaces/account.interface'

const mockAccount: IAccount = {
  avatar: {
    gravatar: {
      hash: '/image'
    },
    tmdb: {
      avatar_path: null
    }
  },
  id: 1234,
  include_adult: true,
  iso_639_1: 'en',
  iso_3166_1: 'en',
  name: '',
  username: 'test/username'
}

export default mockAccount
