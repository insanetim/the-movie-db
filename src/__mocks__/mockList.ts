import { IResponse } from 'src/interfaces/global.interface'
import { IList, IListDetails } from 'src/interfaces/list.interface'

import { mockMovie } from './mockMovie'

const mockList: IList = {
  description: 'test/description',
  id: 1234,
  name: 'test/list',
}

const mockListDetails: IListDetails = {
  ...mockList,
  items: [mockMovie],
  page: 1,
  total_pages: 1,
  total_results: 1,
}

const mockListsResponse: IResponse<IList> = {
  page: 1,
  results: [mockList],
  total_pages: 1,
  total_results: 1,
}

export { mockList, mockListDetails, mockListsResponse }
