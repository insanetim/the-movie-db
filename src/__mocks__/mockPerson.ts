import {
  IExternalIds,
  IPersonCredit,
  IPersonDetails,
} from 'src/interfaces/person.interface'
import { ICredit } from 'src/pages/Credits/types'

const mockExternalIds: IExternalIds = {
  facebook_id: 'facebook_user',
  instagram_id: 'instagram_user',
  tiktok_id: 'tiktok_user',
  twitter_id: 'twitter_user',
}

const mockCreditCast: IPersonCredit[] = [
  {
    character: 'Neo',
    id: 1234,
    popularity: 100500,
    poster_path: '/image',
    release_date: '1999-01-01',
    title: 'The Matrix',
  },
  {
    character: '',
    id: 1223,
    popularity: 100500,
    poster_path: '/image',
    release_date: '2003-01-01',
    title: 'The Matrix Reloaded',
  },
]

const mockedCreditCrew: IPersonCredit[] = [
  {
    department: 'Directing',
    id: 1234,
    job: 'Director',
    popularity: 100500,
    poster_path: '/image',
    release_date: '1999-01-01',
    title: 'The Matrix',
  },
  {
    department: 'Directing',
    id: 4321,
    job: 'Director',
    popularity: 999,
    poster_path: null,
    release_date: '',
    title: 'Jupiter Ascending',
  },
]

const mockedCredits: ICredit[] = [
  {
    key: '1234-the-chosen-one',
    movieSlug: '1234-test-movie',
    posterPath: '/image',
    releaseDate: '1999-01-01',
    releaseDateTitle: '1999',
    role: 'the-chosen-one',
    title: 'the-matrix',
  },
  {
    key: '4321-darth-maul',
    movieSlug: '4321-test-movie',
    posterPath: null,
    releaseDate: '2020-01-01',
    releaseDateTitle: '2020',
    role: 'darth-maul',
    title: 'star-wars',
  },
]

const mockPersonDetails: IPersonDetails = {
  biography: 'test-biography',
  birthday: '1999-01-01',
  external_ids: mockExternalIds,
  gender: 0,
  id: 1234,
  known_for_department: 'Acting',
  movie_credits: {
    cast: mockCreditCast,
    crew: mockedCreditCrew,
  },
  name: 'test-person',
  place_of_birth: 'New York, USA',
  popularity: 100500,
  profile_path: '/image',
}

export { mockPersonDetails, mockedCredits }
