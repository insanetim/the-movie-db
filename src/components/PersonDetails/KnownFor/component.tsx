import { Col, Row, Typography } from 'antd'
import MovieItem from 'src/components/Movies/MovieItem'

import useContainer from './hook'
import { KnownForProps } from './types'

const KnownFor: React.FC<KnownForProps> = ({ credits, department }) => {
  const { sortedItems } = useContainer({ credits, department })

  return (
    <>
      <Col
        className='top-margin'
        span={24}
      >
        <Typography.Title level={2}>Known For</Typography.Title>
      </Col>
      <Col span={24}>
        <Row gutter={[24, 16]}>
          {sortedItems.map(item => {
            return (
              <Col
                key={item.id}
                lg={8}
                md={8}
                sm={12}
                span={24}
                xl={6}
              >
                <MovieItem
                  id={item.id}
                  posterPath={item.poster_path}
                  title={item.title}
                />
              </Col>
            )
          })}
        </Row>
      </Col>
    </>
  )
}

export default KnownFor
