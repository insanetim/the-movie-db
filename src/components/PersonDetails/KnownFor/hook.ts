import { sort } from 'ramda'

import { KnownForHookProps } from './types'

const useContainer = ({
  credits: { cast, crew },
  department,
}: KnownForHookProps) => {
  const items =
    department === 'Acting'
      ? cast
      : crew.filter(item => item.department === department)
  const sortedItems = sort((a, b) => b.popularity - a.popularity, items).slice(
    0,
    12
  )

  return { sortedItems }
}

export default useContainer
