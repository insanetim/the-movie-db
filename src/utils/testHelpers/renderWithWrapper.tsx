import { render, renderHook, RenderOptions } from '@testing-library/react'
import { PropsWithChildren, ReactNode, Suspense } from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import { AppStore, RootState, setupStore } from 'src/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: RootState
  store?: AppStore
}

const createWrapper =
  (store: AppStore) =>
  ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <Suspense fallback={null}>
        <BrowserRouter>{children}</BrowserRouter>
      </Suspense>
    </Provider>
  )

export const renderWithWrapper = (
  ui: ReactNode,
  {
    preloadedState = {} as RootState,
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = createWrapper(store)
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

export const renderHookWithWrapper = <T,>(
  hook: () => T,
  {
    preloadedState = {} as RootState,
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = createWrapper(store)
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
