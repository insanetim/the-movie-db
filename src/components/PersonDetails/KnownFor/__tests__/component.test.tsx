import { mockPersonDetails } from 'src/__mocks__/mockPerson'
import renderWithWrapper from 'src/utils/testHelpers/renderWithWrapper'

import KnownFor from '../component'

describe('KnownFor component', () => {
  it('should match snapshot', () => {
    const { asFragment } = renderWithWrapper(
      <KnownFor
        credits={mockPersonDetails.movie_credits}
        knownForDepartment='Acting'
      />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
