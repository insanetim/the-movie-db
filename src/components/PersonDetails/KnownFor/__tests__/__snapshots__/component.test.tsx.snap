// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KnownFor component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-col ant-col-24 top-margin css-dev-only-do-not-override-1uq9j6g"
  >
    <h2
      class="ant-typography css-dev-only-do-not-override-1uq9j6g"
    >
      Known For
    </h2>
  </div>
  <div
    class="ant-col ant-col-24 css-dev-only-do-not-override-1uq9j6g"
  >
    <div
      class="ant-row css-dev-only-do-not-override-1uq9j6g"
      style="margin-left: -12px; margin-right: -12px; row-gap: 16px;"
    >
      <div
        class="ant-col ant-col-24 ant-col-sm-12 ant-col-md-8 ant-col-lg-8 ant-col-xl-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 12px; padding-right: 12px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-hoverable css-dev-only-do-not-override-1uq9j6g"
          data-testid="movieItemCard"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="The Matrix"
              height="750"
              loading="lazy"
              src="https://image.tmdb.org/t/p/w500/image"
              width="500"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <div
              class="ant-card-meta"
            >
              <div
                class="ant-card-meta-detail"
              >
                <div
                  class="ant-card-meta-title"
                >
                  The Matrix
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-24 ant-col-sm-12 ant-col-md-8 ant-col-lg-8 ant-col-xl-6 css-dev-only-do-not-override-1uq9j6g"
        style="padding-left: 12px; padding-right: 12px;"
      >
        <div
          class="ant-card ant-card-bordered ant-card-hoverable css-dev-only-do-not-override-1uq9j6g"
          data-testid="movieItemCard"
        >
          <div
            class="ant-card-cover"
          >
            <img
              alt="The Matrix Reloaded"
              height="750"
              loading="lazy"
              src="https://image.tmdb.org/t/p/w500/image"
              width="500"
            />
          </div>
          <div
            class="ant-card-body"
          >
            <div
              class="ant-card-meta"
            >
              <div
                class="ant-card-meta-detail"
              >
                <div
                  class="ant-card-meta-title"
                >
                  The Matrix Reloaded
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
